feat: compressed list

dranikpg opened the pull request:
This PR adds a new core type: c compressed sorted list

It will be used to store doc ids in indices and will reduce memory usage for large data sets by a factor of two or three on average. However its iteration speed is much slower...

dranikpg commented on src/core/search/compressed_list.cc:
```diff
@@ -0,0 +1,210 @@
+  return diffs_.size();
+}
+
+// Encode with simple MSB (Most significant bit) encoding: 7 bits of value + 1 to indicate next byte
+absl::Span<uint8_t> CompressedList::WriteVarLen(uint32_t value, absl::Span<uint8_t> buf) {
+  size_t i = 0;
+  do {
+    uint8_t byte = value & 0x7F;
+    value >>= 7;
+    if (value != 0)  // If its not the last byte, set continuation bit
+      byte |= 0x80;
+    buf[i++] = byte;
+  } while (value != 0);
+  return buf.subspan(0, i);
+}
```
I didn't find that many interesting solutions and algorithms on the internet. I want it to be more efficient, because running this loop for decoding a single element might be a little slow. 

Sure, we can unroll it manually with ifs. There is also some code in RDB that uses a similar variable length encoding (though its a little more wasteful, it has 1/2/4 bytes gradation IIRC)

chakaz replied:
How about doing something similar to UTF8, in that the header will have a number of `1`s as the size (in bytes) of the serialized data minus 1?
For example, a number that takes 1 byte will be encoded `0.......` (where `.` are data bits), a number taking 2 bytes will be encoded as `10...... ........`, a number taking 3 bytes will be encoded as `110..... ........ ........` etc
(this isn't identical to UTF8, as it spends fewer bits on encoding)
It should "waste" the same number of bits (as non data representing bits), but this way you can tell how many bits to read by examining the bits byte, and also parsing is easier (just zero out the header)
Just a thought..

royjacobson replied:
relevant - 
https://lemire.me/blog/2017/09/27/stream-vbyte-breaking-new-speed-records-for-integer-compression/

dranikpg replied:
Yes, I'll try this. It should also be possible to implement this without any loops

royjacobson replied:
Did you see they have a library with SIMD support? It's licensed Apache 2.0 (which I _think_ we can use)

chakaz commented on src/core/search/compressed_list.h:
```diff
@@ -0,0 +1,107 @@
+  Iterator begin() const;
+  Iterator end() const;
+
+  void Insert(uint32_t value);  // Insert arbitrary element, needs to scan while list
+  void Remove(uint32_t value);  // Remove arbitrary element, needs to scan while list
```
while -> whole

chakaz commented on src/core/search/compressed_list.h:
```diff
@@ -0,0 +1,107 @@
+  // Find EntryLocation of first entry that is not less than value (std::lower_bound)
+  EntryLocation LowerBound(uint32_t value) const;
+
+  // Push back difference without any decoding. Used only for efficient contruction from sorted list
```
construction (typo)

chakaz commented on src/core/search/compressed_list.h:
```diff
@@ -0,0 +1,107 @@
+  // Push back difference without any decoding. Used only for efficient contruction from sorted list
+  void PushBackDiff(uint32_t diff);
+
+  // Encode interger with variable length encoding into buf and return written subspan
```
integer (typo)

chakaz commented on src/core/search/compressed_list.cc:
```diff
@@ -0,0 +1,210 @@
+
+using namespace std;
+
+CompressedList::Iterator::Iterator(const CompressedList& list) : stash_{}, diffs_{list.diffs_} {
```
I'd assign `stash_` and `diffs_` a default value (alongside their declaration) and have the default constructor `= default` and here only assign `diffs_`. I think it's safer that way for future changes.

chakaz commented on src/core/search/compressed_list.cc:
```diff
@@ -0,0 +1,210 @@
+void CompressedList::Iterator::ReadNext() {
+  if (diffs_.empty()) {
+    diffs_ = {nullptr, 0};
+    return;
```
Should you also assign `nullopt` to `stash_` here?

dranikpg replied:
Theoretically it doesn't matter because reading from an iterator that is equal to end should be forbidden. But I'll add it for better safety

chakaz commented on src/core/search/compressed_list.cc:
```diff
@@ -0,0 +1,210 @@
+
+// Simply encode difference and add to end of diffs array
+void CompressedList::PushBackDiff(uint32_t diff) {
+  array<uint8_t, 16> buf;
```
Could you add a `typedef` for this as well? Something along the lines of `TMaxNumberBuffer`?

chakaz commented on src/core/search/compressed_list.cc:
```diff
@@ -0,0 +1,210 @@
+  uint32_t entry = 0, prev_entry = 0;
+  absl::Span<const uint8_t> diffs_tail{diffs_}, entry_span{};
+
+  while (entry < value && !diffs_tail.empty()) {
+    auto [diff, read] = ReadVarLen(diffs_tail);
+
+    prev_entry = entry;
+    entry += diff;
+
+    entry_span = diffs_tail.subspan(0, read);
+    diffs_tail.remove_prefix(read);
+  }
```
Could you use your iterators here instead? They are much cooler :)

dranikpg replied:
I'll try. The issue is that I want to keep the iterator simple, so I need to calculate the `entry_span` manually with dirty tricks

chakaz commented on src/core/search/compressed_list.cc:
```diff
@@ -0,0 +1,210 @@
+void CompressedList::Insert(uint32_t value) {
+  auto bound = LowerBound(value);
+
+  // At least one element was read and it's equal to value: return to avoid duplicate
```
As such, is this a set then instead of a list?

dranikpg replied:
Yes. Maybe I should call it like this. The name lists just emphasizes that its stored as a contiguous list

chakaz commented on src/core/search/compressed_list.cc:
```diff
@@ -0,0 +1,210 @@
+  auto bound = LowerBound(value);
+
+  // At least one element was read and it's equal to value: return to avoid duplicate
+  if (bound.value == value && !bound.diff_span.empty())
```
Shouldn't checking for `bound`'s `empty()`ness be before comparing the values?

dranikpg replied:
I can, doesn't matter because value is 0 by default

chakaz commented on src/core/search/compressed_list.cc:
```diff
@@ -0,0 +1,210 @@
+  // Now the list certainly contains the bound B > V and possibly A < V (or 0 by default),
+  // so we need to encode both differences diff1 and diff2
+  DCHECK_GT(bound.value, value);
+  DCHECK_LT(bound.prev_value, value);
```
Should this be `DCHECK_LE()` in the case both are `0`?

dranikpg replied:
Good catch. Fixed and added a testcase

chakaz commented on src/core/search/compressed_list.cc:
```diff
@@ -0,0 +1,210 @@
+  size_--;
+
+  // Calculate offset where values diff is stored and determine diffs tail
+  ptrdiff_t diff_offset = bound.diff_span.data() - diffs_.data();
```
Isn't this necessarily `bound.diff_span.size()`?
If so, maybe add a `DCHECK()` or just use that instead, up to you

dranikpg replied:
no, diff_span covers only the small part that encodes the difference of value, not the whole prefix

chakaz commented on src/core/search/compressed_list.cc:
```diff
@@ -0,0 +1,210 @@
+
+  // Calculate offset where values diff is stored and determine diffs tail
+  ptrdiff_t diff_offset = bound.diff_span.data() - diffs_.data();
+  auto diffs_tail = absl::MakeSpan(diffs_).subspan(diff_offset + bound.diff_span.size());
```
I don't understand this line.
You create a subspan starting at the diff? Shouldn't it be `bound.diff_span.data() + diff_offset`?

dranikpg replied:
Answered above, diff_span stores only the small section that makes up the differences of the found element. I'll add a comment

So if we store: `3 5 9 15`
then the diffs will be: `3 2 4 6`
and if we look up 4, then value = 5, prev_value = 3, diff_span = {2} (what element(s) encodes 3->5), so the tail will be 4 and further (`4 6`)

chakaz commented on src/core/search/compressed_list.cc:
```diff
@@ -0,0 +1,210 @@
+  ptrdiff_t diff_offset = bound.diff_span.data() - diffs_.data();
+  auto diffs_tail = absl::MakeSpan(diffs_).subspan(diff_offset + bound.diff_span.size());
+
+  // If its stored at the end, simply truncate it away
```
its -> it's

chakaz reviewed (commented):
Please see my comments, some may be stupid :)
Admittedly, I have not yet looked at the tests

dranikpg commented on src/core/search/compressed_list.h:
```diff
@@ -0,0 +1,107 @@
+    Iterator(const CompressedList& list);
+    Iterator();
+
+    uint32_t operator*() const;
```
True, I'll do it. I just need then to include some common header to already use the existing DocId typedef

dranikpg commented on src/core/search/compressed_list.h:
```diff
@@ -0,0 +1,107 @@
+  };
+
+  // Output iterator to build compressed list from sorted range
+  struct SortedBackInserter {
```
Its a misleading naming... Std has the back_inserter than can be used as a output iterator, so I first called it BackInserter. But it only works with sorted input, so its a sorted back inserter 🙂 Not sure how else to call it

kostasrim reviewed (commented):
Will go over it in more detail later today :)

dranikpg commented on src/core/search/compressed_list.h:
```diff
@@ -0,0 +1,107 @@
+  struct Iterator {
+    friend class CompressedList;
+
+    // To make it work with std container contructors
```
This as well, but the vector constructor (template It, template It) doesn't work without all of those typedefs

dranikpg commented:
Changes:

1. Fixed the small issues and comments
2. Changed the lower bound to work with the iterator. This requires storing the last read span inside the iterator (makes it bigger 🙄 )
3. I changed the encoding to store all the bytes it needs to read ahead. Still need to check how to read them correctly. The link Roy send it cool, but its streamed for decoding large arrays, i do it step by step on iterator. Maybe I will use some kind of decoding buffer in the future to make use of simd, but not in this pr

chakaz commented on src/core/search/compressed_list.cc:
```diff
@@ -0,0 +1,213 @@
+}
+
+absl::Span<uint8_t> CompressedList::WriteVarLen(IntType value, absl::Span<uint8_t> buf) {
+  buf[0] = (value & 0b11111) << 3;
```
I see that you're using a constant 3-bit header here. That's an interesting tradeoff. I guess that it's both simpler to implement, and also saves some bytes for large numbers. However if you expect many diffs in the range of `[2^5, 2^7)` and `[2^13, 2^15)` that would be less efficient.
Your guess is better than mine, so whatever you choose, it's your call.

However, I do think it would be more readable to add some
```cpp
constexpr unsigned kHeaderSize = 3;
```
And use that, as well as `8 - kHeaderSize`.
Perhaps also `0b11111` could be written as `0b1'11'11`
Also, side note, it's more common to use the leftmost bits for the non-data bits, but your call :)

dranikpg replied:
1. Yes, but we'll have to tune the decoding later either way, I'd better not spend many more iterations on compression if it will change in the future, what's here is comparable to what we had
2. Yes, will add to constexpr
3. Yes, storing non data in prefix is better. Will swap them

chakaz reviewed (commented):
Some more comments :)
Please do not think that the number of comments mean anything. Overall I think this is a great class and very nice implementation!

dranikpg commented on src/core/search/compressed_list.h:
```diff
@@ -0,0 +1,113 @@
+  Iterator begin() const;
+  Iterator end() const;
+
+  void Insert(IntType value);  // Insert arbitrary element, needs to scan whole list
+  void Remove(IntType value);  // Remove arbitrary element, needs to scan whole list
```
The correct way to make traversal faster is to use skip lists, I will add it in the future, its out of scope now, because its complicated

dranikpg commented on src/core/search/compressed_list.h:
```diff
@@ -0,0 +1,113 @@
+// A list of sorted unique integers with reduced memory usage.
+// Only differences between successive elements are stored
+// in a variable length encoding.
+class CompressedList {
```
I forgot 😓  yes, I'll rename it

dranikpg commented on src/core/search/compressed_list.cc:
```diff
@@ -0,0 +1,213 @@
+
+namespace {
+
+const int kMaxBufferSize = sizeof(CompressedList::IntType) * 2;
```
const integral values are implicitly constexpr, I'll change it

dranikpg commented on src/core/search/compressed_list.cc:
```diff
@@ -0,0 +1,213 @@
+    it = next_it;
+  }
+
+  return EntryLocation{it.stash_.value_or(0), prev_it.stash_.value_or(0), it.last_read_};
```
Will do, but please not that named aggregate initialization is only part of C++ 20 (but supported by gcc & clang for a long time)

dranikpg commented on src/core/search/compressed_list.cc:
```diff
@@ -0,0 +1,213 @@
+  auto it = begin(), prev_it = end(), next_it = end();
+  while (it != end()) {
+    next_it = it;
+    if (*it >= value || ++next_it == end())
```
Yes, the standard way is to return an iterator to end, but I can't read backwards, so I cap it

dranikpg commented on src/core/search/compressed_list.cc:
```diff
@@ -0,0 +1,213 @@
+
+  // Value is bigger than any other (or list is empty): append required diff at the end
+  if (value > bound.value || bound.diff_span.empty()) {
+    PushBackDiff(value - bound.value);
```
True, need to be cautious. I'm not sure whether to use size at all, I will need in the future as an estimation to select the smaller list for merging

romange commented on src/core/search/compressed_sorted_set.cc:
```diff
@@ -0,0 +1,223 @@
+}
+
+std::pair<CompressedSortedSet::IntType, size_t> CompressedSortedSet::ReadVarLen(
+    absl::Span<const uint8_t> source) {
```
I suggest using flit encoding from helio (see flit.h).
1. It's more efficient in space  for small integers (requires 1 byte for 0-127)
2. It's more efficient with CPU - it does not have loops and a very small number of branches.

see `flit::Encode32` and `flit::Parse64Safe`

romange replied:
in this case you can even use `flit::Parse64Fast` - there is no risk of the vector not being 8-byte aligned.

chakaz replied:
haha, this is exactly the encoding I proposed earlier :)
1 question though: when you say "8-byte aligned" - you don't mean memory alignment, right? Because the vector has integers (potentially) starting with all 8-byte alignment options. I guess you meant that there's always going to be 8 more bytes to read, but that's not necessarily true when reading the last part of the array, right?

dranikpg replied:
Wow, didn't know we have it. Great, its even not used anywhere in DF so far

romange replied:
@chakaz  it's not google varint encoding, it's something else. what have you proposed?

I actually meant that with an allocated byte array, it will always be 8-byte aligned so if for example, we have only a single byte in the array, the process can still access 8 bytes starting from that pointer (yes, 7 others are garbage) without risking to segfault. Parse64Fast has a potential risk of segfaulting because it always loads 8 bytes before parsing.

chakaz replied:
@romange I realize this is not Google varint, I proposed an encoding very similar to this (but bits are slightly flipped), see https://github.com/dragonflydb/dragonfly/pull/1474#discussion_r1241122264

> it will always be 8-byte aligned
Oh, I thought that you were talking about memory alignment (like data structures), sorry

> we have only a single byte in the array, the process can still access 8 bytes starting from that pointer (yes, 7 others are garbage)
What fulfills this guarantee? The data is allocated in a `vector` on the heap

dranikpg commented on src/core/search/compressed_sorted_set.cc:
```diff
@@ -0,0 +1,213 @@
+// The leftmost three bits of the first byte store the number of additional bytes. All following
+// bits store the number itself.
+absl::Span<uint8_t> CompressedSortedSet::WriteVarLen(IntType value, absl::Span<uint8_t> buf) {
+  // TODO: fix flit encoding of large numbers
+  size_t written = base::flit::EncodeT(static_cast<uint64_t>(value), buf.data());
+  return buf.first(written);
```
There is a bug with uint32_t values close to their max, so I cast them for now, needs to be fixed in helio

dranikpg commented:
Changed to flint

chakaz commented on src/core/search/compressed_sorted_set_test.cc:
```diff
@@ -178,12 +178,24 @@ TEST_F(CompressedListTest, BasicRemoveLargeValues) {
-  list.Insert(100500100);
-  list.Insert(212312312);
+  for (int shift = 3; shift < 30; shift++) {
+    uint32_t value = 1u << shift;
```
I would like to see some more tests here, but I guess it's not critical

chakaz reviewed (approved):
Awesome job!
